src/Features/Core/Portable/DesignerAttributes/AbstractDesignerAttributeService.cs

```diff
@@ -2,17 +2,24 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Collections.Generic;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
 using Microsoft.CodeAnalysis.Remote;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.DesignerAttributes
 {
-    internal abstract class AbstractDesignerAttributeService : IDesignerAttributeService
+    internal abstract class AbstractDesignerAttributeService<
+        TCompilationUnitSyntax,
+        TNamespaceDeclarationSyntax,
+        TClassDeclarationSyntax>
+        : IDesignerAttributeService
+        where TCompilationUnitSyntax : SyntaxNode
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TClassDeclarationSyntax : SyntaxNode
```

Comment:
:(

Reply:
?

Reply:
I think adding 3 generic parameters is overkill.

Reply:
this really is hte pattern we use.  when the feature is identical for each language (modulo jsut their syntax nodes being different, we abstract simply with generics in this fashoin).  Then we can share the algo in the common location, and teh derived classes just serve to specify which lang nodes they're using.  

it's effectively a way to have ASTs (but at top level).

Reply:
I get it but the question is about granularity. Yes, we can have a bunch of virtual methods and generic parameters to abstract away a common while loop. But that is imo not worth it. Abstracting away a complex algorithm, or hard to get right code is definitely worth it.

Reply:
so my "worth it" point is smply: would i have to update multiple locations if we decided to change something.  in teh IDE we try to avoid that as much as we can for VB/C# features  (because it just means more stuff to track down and ensure consistency over.  You'll notice that probably about 80% of my latest PRs can just tweak code in a common location and not have to do anything special per language (even though the tweak itself would be very simple in each langauge).

This is really nice and is our common pattern now.  This code here just predates that (primarily because it existed at a time when we didn't even have common nodes) :)

Reply:
I guess I don't like obscure virtual methods like `HasAttributesOrBaseTypeOrIsPartial`. The code is not really language agnostic, it just pretends to be. In VB it returns always true. Shouldn't it actually return true only if the type has attributes, or has a base type, or is partial?

Reply:
Also, it seems that the code is silently assuming that VB modules don't need to be considered.